sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/StateSpec.java

```diff
@@ -36,4 +37,10 @@
    * Use the {@code binder} to create an instance of {@code StateT} appropriate for this address.
    */
   StateT bind(String id, StateBinder<? extends K> binder);
+
+  /**
+   * Given {code coders} are inferred from type arguments defined for this class.
+   * @param coders Array of coders indexed by the type arguments order.
+   */
+  void offerCoders(Coder[] coders);
```

Comment:
So, as I saw it, coders that are already set take precedence, which is why implementations of this method check for coder references which are `null` coders and only set them. This is why I named the method `offerCoders` and not something like `setCoders`. I can improve the javadoc to explain this (Is there a better name for this method?).

If some coders cannot be provided and the coder is needed, the client which uses the coder will throw an exception notifying the user that a coder should have been set but was not. The user should then change their code to specify a coder explicitly using the `StateSpec` factory method which takes `Coder` arguments.
I'm not sure all clients will throw an exception which explains this (I made sure that the exceptions I added to direct runner explain why you are receiving this exception) but I think this holds true for many references to coders across the project.